File: frontend_web/src/pages/auth/signupWizardHelpers.js
import { getPasswordStrength, resolveUploadedImageUrl } from "../../components/auth/authHelpers";
import {
  getDuplicateEmailMessages,
  validateBio,
  validateLicenseNumber,
  validateSpecialization,
  validateWizardForSubmission,
  validateYearsOfExperience,
} from "../../components/auth/authLocalization";

export {
  validateBio,
  validateLicenseNumber,
  validateSpecialization,
  validateWizardForSubmission,
  validateYearsOfExperience,
};

export const EMPTY_SPECIALIST_PROFILE = {
  specialization: "",
  license_number: "",
  years_of_experience: null,
  bio: "",
};

export function revokeWizardProfileImagePreview(preview) {
  if (typeof preview === "string" && preview.startsWith("blob:")) {
    URL.revokeObjectURL(preview);
  }
}

export function getDisplayProfileImage(wizardData) {
  if (wizardData?.profile_image_url) {
    return resolveUploadedImageUrl(wizardData.profile_image_url);
  }

  if (wizardData?.profileImagePreview) {
    return wizardData.profileImagePreview;
  }

  return null;
}

export function getClearWizardProfileImagePatch() {
  return {
    profileImageFile: null,
    profileImagePreview: null,
    profile_image_url: null,
  };
}
export function getStepAfterPersonalInfo(role) {
  return role === "specialist" ? 3 : 4;
}

export function getStepBeforeSecurity(role) {
  return role === "specialist" ? 3 : 2;
}

export function getReviewStep() {
  return 5;
}

export function getSecurityStep() {
  return 4;
}

export const DUPLICATE_EMAIL_TOAST_MESSAGE = "This email is already registered.";
export const DUPLICATE_EMAIL_INLINE_MESSAGE =
  "This email is already registered. Please use a different email or sign in.";

export function getDuplicateEmailCopy(t) {
  return getDuplicateEmailMessages(t);
}

export function isDuplicateEmailError(message) {
  if (typeof message !== "string") {
    return false;
  }

  return /email already exists|already registered/i.test(message);
}

export function isPasswordValid(password) {
  return getPasswordStrength(password).isStrong;
}

export function passwordsMatch(password, confirmPassword) {
  return confirmPassword.length > 0 && password === confirmPassword;
}

export function buildRegistrationPayload(wizardData) {
  const payload = {
    full_name: wizardData.full_name.trim(),
    email: wizardData.email.trim(),
    password: wizardData.password,
    phone: wizardData.phone.trim(),
    role: wizardData.role,
    profile_image_url: wizardData.profile_image_url || null,
  };

  if (wizardData.role === "specialist" && wizardData.specialist_profile) {
    payload.specialist_profile = {
      specialization: wizardData.specialist_profile.specialization.trim(),
      license_number: wizardData.specialist_profile.license_number.trim(),
      years_of_experience: Number(wizardData.specialist_profile.years_of_experience),
      bio: wizardData.specialist_profile.bio?.trim() || null,
    };
  }

  return payload;
}

File: mobile_app/lib/features/auth/utils/signup_wizard_helpers.dart
import '../models/signup_wizard_models.dart';
import '../utils/password_strength.dart';

const int signupWizardTotalSteps = 5;
const int signupSecurityStep = 4;
const int signupReviewStep = 5;

const emptySpecialistProfile = SpecialistProfileData();

int getStepAfterPersonalInfo(SignupRole? role) {
  return role == SignupRole.specialist ? 3 : signupSecurityStep;
}

int getStepBeforeSecurity(SignupRole? role) {
  return role == SignupRole.specialist ? 3 : 2;
}

bool isEmailValid(String email) {
  return RegExp(r'^[^\s@]+@[^\s@]+\.[^\s@]+$').hasMatch(email.trim());
}

bool isFullNameValid(String fullName) {
  return fullName.trim().length >= 2;
}

bool isPhoneValid(String phone) {
  final digitsOnly = phone.replaceAll(RegExp(r'\D'), '');
  return digitsOnly.length >= 7;
}

bool isPasswordValid(String password) {
  return evaluateAuthPasswordStrength(password).isStrong;
}

bool passwordsMatch(String password, String confirmPassword) {
  return confirmPassword.isNotEmpty && password == confirmPassword;
}

SpecialistValidationResult validateSpecialization(String value) {
  final trimmed = value.trim();
  if (trimmed.isEmpty) {
    return const SpecialistValidationResult(
      valid: false,
      message: 'Specialization is required.',
    );
  }
  if (trimmed.length > 150) {
    return const SpecialistValidationResult(
      valid: false,
      message: 'Specialization must not exceed 150 characters.',
    );
  }
  return SpecialistValidationResult(valid: true, value: trimmed);
}

SpecialistValidationResult validateLicenseNumber(String value) {
  final trimmed = value.trim();
  if (trimmed.isEmpty) {
    return const SpecialistValidationResult(
      valid: false,
      message: 'License number is required.',
    );
  }
  if (trimmed.length > 100) {
    return const SpecialistValidationResult(
      valid: false,
      message: 'License number must not exceed 100 characters.',
    );
  }
  return SpecialistValidationResult(valid: true, value: trimmed);
}

SpecialistValidationResult validateYearsOfExperience(String value) {
  final trimmed = value.trim();
  if (trimmed.isEmpty) {
    return const SpecialistValidationResult(
      valid: false,
      message: 'Years of experience is required.',
    );
  }

  final numericValue = int.tryParse(trimmed);
  if (numericValue == null) {
    return const SpecialistValidationResult(
      valid: false,
      message: 'Years of experience is required.',
    );
  }
  if (numericValue < 0) {
    return const SpecialistValidationResult(
      valid: false,
      message: 'Years of experience must be at least 0.',
    );
  }
  return SpecialistValidationResult(valid: true, value: trimmed);
}

SpecialistValidationResult validateBio(String value) {
  if (value.length > 500) {
    return const SpecialistValidationResult(
      valid: false,
      message: 'Bio must not exceed 500 characters.',
    );
  }
  return SpecialistValidationResult(valid: true, value: value);
}

Map<String, dynamic> buildRegistrationPayload({
  required SignupRole role,
  required String fullName,
  required String email,
  required String password,
  required String phone,
  String? profileImageUrl,
  SpecialistProfileData? specialistProfile,
}) {
  final payload = <String, dynamic>{
    'full_name': fullName.trim(),
    'email': email.trim(),
    'password': password,
    'phone': phone.trim(),
    'role': role == SignupRole.parent ? 'parent' : 'specialist',
    if (profileImageUrl != null && profileImageUrl.trim().isNotEmpty)
      'profile_image_url': profileImageUrl.trim(),
  };

  if (role == SignupRole.specialist && specialistProfile != null) {
    payload['specialist_profile'] = specialistProfile.toApiMap();
  }

  return payload;
}

String formatRoleLabel(SignupRole? role) {
  return switch (role) {
    SignupRole.parent => 'Parent',
    SignupRole.specialist => 'Specialist',
    null => '—',
  };
}

String formatExperience(String value) {
  final trimmed = value.trim();
  if (trimmed.isEmpty) {
    return '—';
  }
  final years = int.tryParse(trimmed);
  if (years == null) {
    return trimmed;
  }
  return '$years year${years == 1 ? '' : 's'}';
}

bool isDuplicateEmailError(String? message) {
  if (message == null || message.isEmpty) {
    return false;
  }
  return RegExp(
    r'email already exists|already registered',
    caseSensitive: false,
  ).hasMatch(message);
}

class SpecialistValidationResult {
  const SpecialistValidationResult({
    required this.valid,
    this.message,
    this.value,
  });

  final bool valid;
  final String? message;
  final String? value;
}
